fbpic/particles/deposition/threading_methods.py

```diff
@@ -0,0 +1,454 @@
+            z_cell = invdz*(zj - zmin) - 0.5
+
+            # Calculate rho
+            # --------------------------------------------
+            # Mode 0
+            R_m0_scal = wj * exptheta_m0
+            # Mode 1
+            R_m1_scal = wj * exptheta_m1
+
+            # Original index of the uppper and lower cell
+            ir_cell = int(math.floor( r_cell ))
+            iz_cell = int(math.floor( z_cell ))
+
+            # Treat the boundary conditions
+            # guard cells in lower r
+            if ir_cell < 0:
+                ir_cell = 0
+            # absorbing in upper r
+            if ir_cell > Nr-1:
+                ir_cell = Nr-1
+            # periodic boundaries in z
+            if iz_cell < 0:
+                iz_cell += Nz
+            if iz_cell > Nz-1:
+                iz_cell -= Nz
+
+            # Boundary Region Shifts
+            ir_flip = int( math.floor(r_cell) )
+
+            R_m0_00 = r_shape_linear(r_cell, 0)*z_shape_linear(z_cell, 0) * R_m0_scal
+            R_m0_01 = r_shape_linear(r_cell, 0)*z_shape_linear(z_cell, 1) * R_m0_scal
+            R_m1_00 = r_shape_linear(r_cell, 0)*z_shape_linear(z_cell, 0) * R_m1_scal
+            R_m1_01 = r_shape_linear(r_cell, 0)*z_shape_linear(z_cell, 1) * R_m1_scal
+
+            if ir_flip == -1:
+                R_m0_00 += r_shape_linear(r_cell, 1)*z_shape_linear(z_cell, 0) * R_m0_scal
+                R_m0_01 += r_shape_linear(r_cell, 1)*z_shape_linear(z_cell, 1) * R_m0_scal
+                R_m1_00 += r_shape_linear(r_cell, 1)*z_shape_linear(z_cell, 0) * R_m1_scal
+                R_m1_01 += r_shape_linear(r_cell, 1)*z_shape_linear(z_cell, 1) * R_m1_scal
+            else:
+                R_m0_10 = r_shape_linear(r_cell, 1)*z_shape_linear(z_cell, 0) * R_m0_scal
+                R_m0_11 = r_shape_linear(r_cell, 1)*z_shape_linear(z_cell, 1) * R_m0_scal
+                R_m1_10 = r_shape_linear(r_cell, 1)*z_shape_linear(z_cell, 0) * R_m1_scal
+                R_m1_11 = r_shape_linear(r_cell, 1)*z_shape_linear(z_cell, 1) * R_m1_scal
+
+            # Cell shifts for the simulation boundaries
+            shift_r = 0
+            shift_z = 0
+            if ir_cell+1 > (Nr-1):
+                shift_r = -1
+            if iz_cell+1 > Nz-1:
+                shift_z -= Nz
+            # Write to thread local arrays
+            rho_m0_thread[iz_cell, ir_cell] += R_m0_00
+            rho_m1_thread[iz_cell, ir_cell] += R_m1_00
+
+            rho_m0_thread[iz_cell+1 + shift_z, ir_cell] += R_m0_01
+            rho_m1_thread[iz_cell+1 + shift_z, ir_cell] += R_m1_01
+
+            rho_m0_thread[iz_cell, ir_cell+1 + shift_r] += R_m0_10
+            rho_m1_thread[iz_cell, ir_cell+1 + shift_r] += R_m1_10
+
+            rho_m0_thread[iz_cell+1 + shift_z, ir_cell+1 + shift_r] += R_m0_11
+            rho_m1_thread[iz_cell+1 + shift_z, ir_cell+1 + shift_r] += R_m1_11
+
+        # Write thread local deposition arrays to global deposition arrays

```

Comment:
Just a random thought: wouldn't the operation below be faster if we initialize rho_m0_global with a different shape (i.e. such that we do `rho_m0_global[tx,:,:] = rho_m0_thread` and thus the copying is contiguous in memory).

I agree that then the reduction operation is then maybe not as fast, but we could also write a numba reduction function that uses prange (with parallelization across the z index for instance).